## Part 1

In [1]:
import math
import functools

In [2]:
data = """
162,817,812
57,618,57
906,360,560
592,479,940
352,342,300
466,668,158
542,29,236
431,825,988
739,650,466
52,470,668
216,146,977
819,987,18
117,168,530
805,96,715
346,949,466
970,615,88
941,993,340
862,61,35
984,92,344
425,690,689
"""

with open("input.txt", "r") as f:
    data = f.read()

In [3]:
def get_positions(data:str):
    positions = []
    for line in data.strip().split("\n"):
        positions.append(tuple([int(n) for n in line.split(",")]))
    return positions

@functools.cache
def euclidean_distance(a, b):
    ax,ay,az = a
    bx,by,bz = b
    return math.sqrt(pow(ax-bx, 2) + pow(ay-by, 2) + pow(az-bz, 2))
    # return math.sqrt(pow(cx - fx, 2) + pow(cy - fy, 2))

    
positions = get_positions(data)

In [4]:
def get_distances(positions):
    distances = []
    for a_idx in range(len(positions)):
        for b_idx in range(a_idx+1, len(positions)):
            a = positions[a_idx]
            b = positions[b_idx]
            distances.append((a, b, euclidean_distance(a,b)))

    return sorted(distances, key=lambda x: x[2])

# get_distances(get_positions(data))

In [5]:
def get_circuits(positions, distances, max_connections=10):
    relations = {}

    for box in positions:
        relations[box] = set()
        relations[box].add(box)

    for a, b, _ in distances[:max_connections]:
        relations[a].add(b)
        relations[b].add(a)

    def walk(box, circuit, relations):
        if box not in relations:
            return

        current = relations.pop(box)
        
        if len(current) == 1:
            circuit.append(list(current)[0])
            return

        for box in current:
            if box in circuit:
                continue
            circuit.append(box)
            walk(box, circuit, relations)

        return

    circuits = []

    for box in positions:
        circuit = []
        walk(box, circuit, relations)
        circuits.append(circuit)

    return list(filter(lambda x: len(x) > 0, circuits))

def calculate_result(circuits):
    return math.prod(sorted([len(circuit) for circuit in circuits], reverse=True)[:3])
    

calculate_result(get_circuits(get_positions(data), get_distances(get_positions(data)), 1000))

52668

Attempts
- 308 your answer is too low
- 5474 your answer is too low
- ⭐️ 52668

## Part 2

In [6]:
def get_circuit_count(mappings):
    relations = {}

    for key,value in mappings.items():
        relations[key] = [*value]

    boxes = mappings.keys()

    circuits = []

    def walk(box, circuit, relations):
        if box not in relations:
            return

        current = relations.pop(box)
        if len(current) == 1:
            circuit.append(current[0])
            return

        for box in current:
            if box in circuit:
                continue
            circuit.append(box)
            walk(box, circuit, relations)

        return

    for box in boxes:
        circuit = []
        walk(box, circuit, relations)
        circuits.append(circuit)

    return len(list(filter(lambda x: len(x) > 0, circuits)))

def connect(positions, distances):
    relations = {}
    
    for box in positions:
        relations[box] = set()
        relations[box].add(box)

    iterations = 0

    for a, b, _ in distances:
        iterations += 1
        relations[a].add(b)
        relations[b].add(a)

        if iterations < 6500:
            continue

        count = get_circuit_count(relations)

        if count <= 1:
            print(f"{a=} {b=}")
            return a[0] * b[0]

connect(get_positions(data), get_distances(get_positions(data)))

a=(37565, 7239, 99565) b=(39240, 1604, 85100)


1474050600

Attempts:
- ⭐️ 1474050600